File: app/src/main/java/com/oguz/spy/ads/RewardedAdManager.kt
package com.oguz.spy.ads

import android.app.Activity
import android.content.Context
import com.google.android.gms.ads.AdError
import com.google.android.gms.ads.AdRequest
import com.google.android.gms.ads.FullScreenContentCallback
import com.google.android.gms.ads.LoadAdError
import com.google.android.gms.ads.rewarded.RewardedAd
import com.google.android.gms.ads.rewarded.RewardedAdLoadCallback
import com.oguz.spy.ads.AdIds.REWARDED_TEST

class RewardedAdManager(
    private val context: Context,
    private val adUnitId: String = REWARDED_TEST // Test ID
) {
    private var rewardedAd: RewardedAd? = null
    private var isLoading = false

    fun loadAd(onAdLoaded: () -> Unit = {}, onAdFailedToLoad: (String) -> Unit = {}) {
        if (isLoading) return

        isLoading = true
        val adRequest = AdRequest.Builder().build()

        RewardedAd.load(
            context,
            adUnitId,
            adRequest,
            object : RewardedAdLoadCallback() {
                override fun onAdLoaded(ad: RewardedAd) {
                    rewardedAd = ad
                    isLoading = false
                    onAdLoaded()
                }

                override fun onAdFailedToLoad(error: LoadAdError) {
                    rewardedAd = null
                    isLoading = false
                    onAdFailedToLoad(error.message)
                }
            }
        )
    }

    fun showAd(
        activity: Activity,
        onUserEarnedReward: (Int, String) -> Unit,
        onAdDismissed: () -> Unit = {},
        onAdShowFailed: (String) -> Unit = {}
    ) {
        if (rewardedAd != null) {
            rewardedAd?.fullScreenContentCallback = object : FullScreenContentCallback() {
                override fun onAdDismissedFullScreenContent() {
                    rewardedAd = null
                    onAdDismissed()
                    // Yeni reklam yükle
                    loadAd()
                }

                override fun onAdFailedToShowFullScreenContent(error: AdError) {
                    rewardedAd = null
                    onAdShowFailed(error.message)
                }

                override fun onAdShowedFullScreenContent() {
                    // Reklam gösterildi
                }
            }

            rewardedAd?.show(activity) { reward ->
                // Kullanıcı ödülü kazandı
                onUserEarnedReward(reward.amount, reward.type)
            }
        } else {
            onAdShowFailed("Reklam henüz yüklenmedi")
            loadAd()
        }
    }

    fun isAdReady(): Boolean = rewardedAd != null
}

File: lib/ads/rewarded_ad_manager.dart
import 'package:google_mobile_ads/google_mobile_ads.dart';
import 'ad_ids.dart';

class RewardedAdManager {
  RewardedAd? _rewardedAd;
  bool _isLoading = false;

  void loadAd({
    Function()? onAdLoaded,
    Function(String)? onAdFailedToLoad,
  }) {
    if (_isLoading) return;
    _isLoading = true;

    RewardedAd.load(
      adUnitId: AdIds.rewarded,
      request: const AdRequest(),
      rewardedAdLoadCallback: RewardedAdLoadCallback(
        onAdLoaded: (ad) {
          _rewardedAd = ad;
          _isLoading = false;
          onAdLoaded?.call();
        },
        onAdFailedToLoad: (error) {
          _rewardedAd = null;
          _isLoading = false;
          onAdFailedToLoad?.call(error.message);
        },
      ),
    );
  }

  void showAd({
    required Function(int amount, String type) onUserEarnedReward,
    Function()? onAdDismissed,
    Function(String)? onAdShowFailed,
  }) {
    if (_rewardedAd == null) {
      onAdShowFailed?.call('Reklam henüz yüklenmedi');
      loadAd();
      return;
    }

    _rewardedAd!.fullScreenContentCallback = FullScreenContentCallback(
      onAdDismissedFullScreenContent: (ad) {
        ad.dispose();
        _rewardedAd = null;
        onAdDismissed?.call();
        loadAd();
      },
      onAdFailedToShowFullScreenContent: (ad, error) {
        ad.dispose();
        _rewardedAd = null;
        onAdShowFailed?.call(error.message);
        loadAd();
      },
    );

    _rewardedAd!.show(
      onUserEarnedReward: (ad, reward) {
        onUserEarnedReward(reward.amount.toInt(), reward.type);
      },
    );
  }

  bool get isAdReady => _rewardedAd != null;

  void dispose() {
    _rewardedAd?.dispose();
    _rewardedAd = null;
  }
}
